### PR TITLE
Rewrite `opening_brace` rule with SwiftSyntax

### PR DESCRIPTION
**Pull request by kishikawakatsumi**
Fix https://github.com/realm/SwiftLint/issues/4610
Fix https://github.com/realm/SwiftLint/issues/5114
Fix https://github.com/realm/SwiftLint/issues/3470
Fix https://github.com/realm/SwiftLint/issues/3574
Fix https://github.com/realm/SwiftLint/issues/2632
Fix https://github.com/realm/SwiftLint/issues/3476
Fix https://github.com/realm/SwiftLint/issues/3756

Reworked `opening_brace` rule using SwiftSyntax.
While only violation detection can be implemented by simply visiting `TokenSyntax`, but it is necessary to visit each `*Syntax` to achieve automatic correction. The reason for this is that the whitespace before the brace is not the trivia of the brace, but the trivia of the token before it. So this is a pretty big change.

### CODE REVIEW

File: Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift [hunk 1]
```diff
@@ -77,9 +18,6 @@ struct OpeningBraceRule: CorrectableRule, ConfigurationProviderRule {
             Example("if let a = b { }"),
             Example("while a == b { }"),
             Example("guard let a = b else { }"),
-            Example("if\n\tlet a = b,\n\tlet c = d\n\twhere a == c\n{ }"),
-            Example("while\n\tlet a = b,\n\tlet c = d\n\twhere a == c\n{ }"),
-            Example("guard\n\tlet a = b,\n\tlet c = d\n\twhere a == c else\n{ }"),
```
Comment (kishikawakatsumi) [inline Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift:82]: Writing a `where` clause to join multiple conditions is no longer permitted in Swift. Therefore, I have removed this test case.

File: Tests/SwiftLintFrameworkTests/OpeningBraceRuleTests.swift [hunk 2]
```diff
@@ -13,9 +13,6 @@ class OpeningBraceRuleTests: SwiftLintTestCase {
             Example("if let a = b { }"),
             Example("while a == b { }"),
             Example("guard let a = b else { }"),
-            Example("if\n\tlet a = b,\n\tlet c = d\n\twhere a == c\n{ }"),
-            Example("while\n\tlet a = b,\n\tlet c = d\n\twhere a == c\n{ }"),
-            Example("guard\n\tlet a = b,\n\tlet c = d\n\twhere a == c else\n{ }"),
```
Comment (kishikawakatsumi) [inline Tests/SwiftLintFrameworkTests/OpeningBraceRuleTests.swift:18]: Writing a `where` clause to join multiple conditions is no longer permitted in Swift. Therefore, I have removed this test case.

**Review by SimplyDanny — CHANGES_REQUESTED** [review]
Thanks for taking that up!

Please add a CHANGELOG entry to let people know about the fixes.

File: Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift [hunk 3]
```diff
@@ -169,73 +185,1051 @@ struct OpeningBraceRule: CorrectableRule, ConfigurationProviderRule {
...
+
+        return nil
+    }
+}
+
+private extension ClosureExprSyntax {
+    func violationPosition() -> AbsolutePosition? {
+        let openingBrace = leftBrace
+
+        if let functionCall = parent?.as(FunctionCallExprSyntax.self) {
```
[the remark this answers is not in the document]
Comment (kishikawakatsumi) [reply Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift:1213]: Fixed by f07f7f54ce145f886b937f794308c087631f23d0

File: Source/SwiftLintBuiltInRules/Rules/Style/PreferSelfInStaticReferencesRule.swift [hunk 4]
```diff
@@ -228,8 +228,7 @@ private class Visitor: ViolationsSyntaxVisitor {
         if let varDecl = node.parent?.parent?.parent?.as(VariableDeclSyntax.self) {
             if varDecl.parent?.is(CodeBlockItemSyntax.self) == true // Local variable declaration
                 || varDecl.bindings.onlyElement?.accessor != nil    // Computed property
-                || !node.type.is(SimpleTypeIdentifierSyntax.self)   // Complex or collection type
```
[the remark this answers is not in the document]
Comment (kishikawakatsumi) [reply Source/SwiftLintBuiltInRules/Rules/Style/PreferSelfInStaticReferencesRule.swift]: Fixed 9c57348e4

File: Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift [hunk 5]
```diff
@@ -169,73 +185,1051 @@ struct OpeningBraceRule: CorrectableRule, ConfigurationProviderRule {
...
+
+            return .visitChildren
+        }
+
+        override func visit(_ node: FunctionDeclSyntax) -> SyntaxVisitorContinueKind {
+            guard let body = node.body else {
+                return .visitChildren
+            }
+
+            var isMultilineFunction: Bool {
```
[the remark this answers is not in the document]
Comment (kishikawakatsumi) [reply Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift]: Fixed 9b2c86ee8

File: Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift [hunk 6]
```diff
@@ -1,67 +1,8 @@
...
-                }
-                bracketCount -= 1
-            }
-        }
-        return nil
-    }
-}
-
-struct OpeningBraceRule: CorrectableRule, ConfigurationProviderRule {
+// swiftlint:disable file_length type_body_length
```
[the remark this answers is not in the document]
Comment (kishikawakatsumi) [reply Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift:4]: Fix f0aa7154a39fe43cbeaf4ea6459a50ddc6e2a937

File: Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift [hunk 7]
```diff
@@ -1,67 +1,8 @@
-import Foundation
-import SourceKittenFramework
+import SwiftSyntax
+import SwiftSyntaxBuilder
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift:2]: Is this import really needed?

File: Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift [hunk 8]
```diff
@@ -169,73 +225,1044 @@ struct OpeningBraceRule: CorrectableRule, ConfigurationProviderRule {
...
+            for a in b
+            ↓{
+
+            }
+            """):
+                Example("""
+            for a in b {
+
+            }
+            """),
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Style/OpeningBraceRule.swift:458]: The common indentation style for corrected examples in SwiftLint would be like:
```suggestion
                Example("""
                for a in b {

                }
                """),
```

**Review by SimplyDanny — COMMENTED** [review]
Thanks for the updates!

A few general comments/questions:

1. The implemented corrections will remove comments. While I don't like that, it might be acceptable for the time being. I see a few issues incoming, though.
2. The list of examples is quite long. We typically move them into a separate file when they become too extensive.
3. The examples are not only tests but they also serve as documentation. We should exclude a few from being documented by adding the `excludeFromDocumentation` parameter, especially when they only test special cases.
4. Did you check the OSS findings? I've seen a few which trigger while they shouldn't or don't trigger anymore while they should.

**Comment by SimplyDanny** [comment]
Rebased, resolved conflicts, adapted code to current state of the project, simplified a little.

What's left:

- [x] Resolve review comments
- [ ] Fix changelog
- [ ] Check OSS findings

**Pull request merged**